pkg/core/resource_graph.go

```diff
@@ -10,58 +12,120 @@ type (
+	return rg.underlying.GetVertex(id)
 }
 
-func (cg *ResourceGraph) GetResource(id string) Resource {
-	return cg.underlying.GetVertex(id)
+func (rg *ResourceGraph) GetDependency(source string, target string) *graph.Edge[Resource] {
+	return rg.underlying.GetEdge(source, target)
 }
 
-func (cg *ResourceGraph) GetDependency(source string, target string) *graph.Edge[Resource] {
-	return cg.underlying.GetEdge(source, target)
+func (rg *ResourceGraph) ListResources() []Resource {
+	return rg.underlying.GetAllVertices()
 }
 
-func (cg *ResourceGraph) ListResources() []Resource {
-	return cg.underlying.GetAllVertices()
+func (rg *ResourceGraph) ListDependencies() []graph.Edge[Resource] {
+	return rg.underlying.GetAllEdges()
 }
 
-func (cg *ResourceGraph) ListDependencies() []graph.Edge[Resource] {
-	return cg.underlying.GetAllEdges()
+func (rg *ResourceGraph) VertexIdsInTopologicalOrder() ([]string, error) {
+	return rg.underlying.VertexIdsInTopologicalOrder()
 }
 
-func (cg *ResourceGraph) VertexIdsInTopologicalOrder() ([]string, error) {
-	return cg.underlying.VertexIdsInTopologicalOrder()
+func (rg *ResourceGraph) GetDownstreamDependencies(source Resource) []graph.Edge[Resource] {
+	return rg.underlying.OutgoingEdges(source)
 }
 
-func (cg *ResourceGraph) GetDownstreamDependencies(source Resource) []graph.Edge[Resource] {
-	return cg.underlying.OutgoingEdges(source)
+func (rg *ResourceGraph) GetDownstreamResources(source Resource) []Resource {
+	return rg.underlying.OutgoingVertices(source)
 }
 
-func (cg *ResourceGraph) GetDownstreamResources(source Resource) []Resource {
-	return cg.underlying.OutgoingVertices(source)
+func (rg *ResourceGraph) GetUpstreamDependencies(source Resource) []graph.Edge[Resource] {
+	return rg.underlying.IncomingEdges(source)
 }
 
-func (cg *ResourceGraph) GetUpstreamDependencies(source Resource) []graph.Edge[Resource] {
-	return cg.underlying.IncomingEdges(source)
+func (rg *ResourceGraph) GetUpstreamResources(source Resource) []Resource {
+	return rg.underlying.IncomingVertices(source)
 }
 
-func (cg *ResourceGraph) GetUpstreamResources(source Resource) []Resource {
-	return cg.underlying.IncomingVertices(source)
+func (rg *ResourceGraph) TopologicalSort() ([]string, error) {
+	return rg.underlying.VertexIdsInTopologicalOrder()
 }
 
-func (cg *ResourceGraph) TopologicalSort() ([]string, error) {
-	return cg.underlying.VertexIdsInTopologicalOrder()
+// AddDependenciesReflect uses reflection to inspect the fields of the resource given
+// and add dependencies for each direct (ie, first-level) dependency.
+//
+// Supported field types (`*T` is a struct that implements Resource)
+// - `SingleDependency   Resource`
+// - `SpecificDependency *T`
+// - `DependencyArray  []Resource`
+// - `SpecificDepArray []*T`
+// - `DependencyMap  map[string]Resource`
+// - `SpecificDepMap map[string]*T`
+func (rg *ResourceGraph) AddDependenciesReflect(source Resource) {
+	rv := reflect.ValueOf(source)
```

Comment:
I find these hypershort names hard to follow through the method. I'd prefer `resVal` etc.